woodwork/tests/utils/test_accessor_utils.py

```diff
@@ -255,7 +255,7 @@ def test_get_invalid_schema_message_index_checks(sample_df):
 
     nan_df = pd.DataFrame(
         {
-            "id": pd.Series([5, 4, None, 2], dtype="float64"),
+            "id": pd.Series([5, 4, 3, 2], dtype="float64"),
```

Comment:
I'm not entirely clear now on what we were testing here, but I think perhaps this test was intended to confirm that if a single null value is present that we don't generate an invalid schema message? @tamargrey Do you recall?

If we were testing that introducing a single null value was ok before, I think maybe with this new behavior of not allowing any nulls in the index, we may need to update `get_invalid_schema_message` to return a message if some operation causes null values to be introduced in the index that invalidates a schema (like was happening in `test_concat_cols_mismatched_index_adds_single_nan` before the change you made to not set the index to `id`). This would be similar to the assertion that precedes this.

Reply:
I don't fully remember, but given the fact that the dataframe is called `nan_df` and that the other two index-related schema vilidity checks were already made in this test, I'd guess that the goals was to confirm that an index column containing nans was acceptable to `get_invalid_schema_message`, in which case the check with this dataframe isn't necessary

Reply:
It sounds like this case is safe to remove then, right?

Reply:
I guess it depends on if we want to handle the situation where a schema is invalid because an index contains a single null value in `get_invalid_schema_message`. The `.is_unique` check will cover a scenario where we have multiple null values, I think, but we wouldn't have anything in that function to catch a single missing value. I'm not sure how important that really is though.

Reply:
If the following works, then we need to add a case to `get_invalid_schema_message` to catch it and raise an error and update this test to confirm that that message is produced 

```
df = pd.DataFrame(
      {
          "id": pd.Series([5, 4, 3, 2], dtype="float64"),
          "col": pd.Series(["b", "b", "b", "d"], dtype="category"),
      }
  )
  df.ww.init(index="id")
  df_schema = df.ww.schema

nan_df = df.replace({3:None})
nan_df.ww.init_with_full_schema(df_schema)

```

Reply:
@tamargrey It works, but we first have to do a couple other things - we need to make sure the dtypes are consistent and set the pandas index:
```
df = pd.DataFrame(
      {
          "id": pd.Series([5, 4, 3, 2], dtype="float64"),
          "col": pd.Series(["b", "b", "b", "d"], dtype="category"),
      }
  )
df.ww.init(index="id")
df_schema = df.ww.schema

nan_df = df.replace({3:None})
nan_df["id"] = nan_df["id"].astype("float64")
nan_df = nan_df.set_index("id", drop=False)
nan_df.ww.init_with_full_schema(df_schema)
```
Without those other two steps we get a dtype mismatch error and a index mismatch error, but once those changes are made, the init works fine without issue.

Reply:
oops, yeah forgot about the other requirements. In that case, we shoud add a check like `if dataframe[schema.index].isnull().any(): return "Index contains null values"`

Reply:
I was looking through the validation process and had thoughts/questions. If I'm understanding correctly, `init_with_full_schema` assigns or replaces the underlying schema whereas `init_with_partial_schema` creates and builds the schema. As a result, these two places (possibly more?) where the same validation (more or less) needs to happen.

Reply:
Would it be possible to take schema in `init_with_full_schema` and convert it back to parameters needed for `init_with_partial_schema`?

Reply:
It'd definitely be possible, and when you pass a full schema into `df.ww.init` that's now what happens. But when you pass parameters into `init_with_partial_schema`, it'll change the dataframe to fit those parameters if it has to.

We specifically kept `init_with_full_schema` with its current behavior that won't make any changes to the dataframe, since it's useful in cases where we expect the dataframe to be valid and wouldn't want the partial schema init to, say, change dtypes or reset the underlying index. We use it a couple of places around woodwork and featuretools and rely on that behavior. 

I definitely agree that having this logic twice over doesn't feel the best, though I don't think we have anywhere where the validation is actually being performed twice in a row. It'd be great if we could have shared utils for this, but the reason we dont, if I'm rmembering correctly, is that we want `get_invalid_schema_message` to return the explanation without yet raising an error.

Reply:
If `init_with_partial_schema` only changes the dataframe when necessary, it sounds like this method shouldn't change the dataframe at all in some cases where it's not necessary. I'm not entirely sure that I understand the need for both, but not against leaving this as is.

Reply:
This is definitely a valid line of questioning. I think there's definitely good reason to have `init_with_partial_schema` when setting `validate=False`--we use that a lot across woodwork when we know the schema will be valid. There's not much usage with `validate=True`, but the utilities that make up the validation are independently important. `get_invalid_schema_message` and `is_schema_valid` are both used by evalml.

So we still need to be able to define what is and is not a valid schema, and with the changes in this PR, we should add a check that a dataframe passed into `get_invalid_schema_message` or `is_schema_valid` does not have any nans in its index column.

Reply:
updated